src/qbe_ir_generator.cpp

```diff
@@ -1084,11 +1084,24 @@ void QbeIrGenerator::Visit(const SimpleAssignmentExprNode& assign_expr) {
   int rhs_num = num_recorder.NumOfPrevExpr();
   if (assign_expr.lhs->type->IsPtr()) {
     // Assign pointer address to another pointer.
-    WriteInstr_("storel {}, {}", FuncScopeTemp{rhs_num},
-                FuncScopeTemp{reg_num_to_id_num.at(lhs_num)});
+    if (assign_expr.lhs->is_global) {
+      WriteInstr_("storel {}, {}", FuncScopeTemp{rhs_num},
+                  user_defined::GlobalPointer{reg_num_to_id.at(lhs_num)});
+    } else {
+      WriteInstr_("storel {}, {}", FuncScopeTemp{rhs_num},
+                  FuncScopeTemp{reg_num_to_id_num.at(lhs_num)});
+    }
   } else {
-    WriteInstr_("storew {}, {}", FuncScopeTemp{rhs_num},
-                FuncScopeTemp{reg_num_to_id_num.at(lhs_num)});
+    // Global array subscripting will return the target address instead of the
+    // address of `id`.
+    if (assign_expr.lhs->is_global &&
+        !dynamic_cast<ArrSubExprNode*>(assign_expr.lhs.get())) {
```

Comment:
The comment refers to subscripting a global array; however, the if block is triggered when `lhs` is a global non-subscripting array. Additionally, could you provide a small example? I'm a bit confused about why `ArrSubExprNode` is mentioned here.

Reply:
I've updated the predicate now.

```
int b = 3
int a[3] = {6, 5, 3};

int main() {
  b = 4;
  a[0] = 7;
  return 0;
}

```

In `ArrSubExprNode`, even if an array is global, the only difference for array subscripting is to get the base address of the array from `id`. The target address is still stored in a temporary register. 

```cpp
 if (arr_sub_expr.is_global) {
    const auto id = reg_num_to_id.at(reg_num);
    WriteInstr_("{} =l add {}, {}", FuncScopeTemp{res_addr_num},
                user_defined::GlobalPointer{id}, FuncScopeTemp{offset});
  } else {
    // address of the first element
    const int base_addr = reg_num_to_id_num.at(reg_num);
    WriteInstr_("{} =l add {}, {}", FuncScopeTemp{res_addr_num},
                FuncScopeTemp{base_addr}, FuncScopeTemp{offset});
  }

  // load value from res_addr
  const int res_num = NextLocalNum();
  WriteInstr_("{} =w loadw {}", FuncScopeTemp{res_num},
              FuncScopeTemp{res_addr_num});
  reg_num_to_id_num[res_num] = res_addr_num;
  num_recorder.Record(res_num);
```

`b` will get its address from `reg_num_to_id` and `a[0]` will get its address from `reg_num_to_id_num`. Sorry for the confusion! Hopefully, this version is more clearer.

```cpp
     if (assign_expr.lhs->is_global) {
      // Global array subscripting will return the target address instead of the
      // address of `id`.
      if (dynamic_cast<ArrSubExprNode*>(assign_expr.lhs.get())) {
        WriteInstr_("storew {}, {}", FuncScopeTemp{rhs_num},
                    FuncScopeTemp{reg_num_to_id_num.at(lhs_num)});
      } else {
        WriteInstr_("storew {}, {}", FuncScopeTemp{rhs_num},
                    user_defined::GlobalPointer{reg_num_to_id.at(lhs_num)});
      }
    } else {
      WriteInstr_("storew {}, {}", FuncScopeTemp{rhs_num},
                  FuncScopeTemp{reg_num_to_id_num.at(lhs_num)});
    }
```

Reply:
Does this kind of issue only happen with `ArrSubExprNode`? I'm concerned that dereferencing other expressions might also require special treatment.

Reply:
> Does this kind of issue only happen with `ArrSubExprNode`? I'm concerned that dereferencing or other expressions might also require special treatment.

It might, we'll fix that if we got there.

Reply:
:ok_hand: Leave this as-is. :love_you_gesture: